posts/inside-rust/2021-08-27-Splitting-const-generics.md

```diff
@@ -0,0 +1,127 @@
+
+Additionally, we have to figure out which types we *even want* to allow as const parameter types. This ties into the discussion
+about ["structural match"](https://github.com/rust-lang/rust/issues/74446), which is still ongoing.
+
+While the issues mentioned above are definitely not trivial, I wouldn't be surprised to see this ready for stabilization in a few months.
+
+### `feature(generic_const_exprs)`
+
+Without any unstable feature, const arguments must either be a fully concrete expression or a generic parameter by itself, so constants like `N + 1` are forbidden. With this feature, expressions using generic parameters are possible. 
+
+```rust
+#![feature(generic_const_exprs)]
+
+fn split_first<T, const N: usize>(arr: [T; N]) -> (T, [T; N - 1]) {
+    // ...
+}
+
+struct BitSet<const SIZE: usize>
+where
+    [u8; (SIZE + 7) / 8]: Sized,
+{
+    storage: [u8; (SIZE + 7) / 8],
+}
+```
+
+We currently require the user to add bounds asserting that generic constants evaluate successfully. For all constants visible in the API of an item, these bounds are added implicitly. 
+
+If the constant expression `expr` of type `Foo` would otherwise not be used in the `where`-clauses or function signature, we add an otherwise irrelevant bound mentioning `expr` to the `where`-clauses of our item. For this one can define a `struct Evaluatable<const N: Foo>;` and use `Evaluatable<{ expr }>:` as a bound. If `expr` is of type `usize` we tend to use `[u8; expr]:`
+or `[u8; expr]: Sized` for this. While it is highly likely that we will add a dedicated syntax for these bounds in the future, we are waiting with this until the rest of this feature is more mature.
+
+This feature is still far from being stable and has some [**major** unsolved issues](https://github.com/rust-lang/project-const-generics/blob/master/design-docs/anon-const-substs.md). Especially for constants inside of `where`-bounds there are a lot of subtle bugs and backwards incompatibilities we have to fix before we can even think about how to stabilize this.
+
+### `feature(const_generics_defaults)`
+
+Similar to type parameter defaults, this feature adds the ability to declare default values for const parameters.
+
+```rust
+#![feature(const_generics_defaults)]
+
+struct ArrayStorage<T, const N: usize = 2> {
+    arr: [T; N],
+}
+
+impl<T> ArrayStorage<T> {
+    fn new(a: T, b: T) -> ArrayStorage<T> {
+        ArrayStorage {
+            arr: [a, b],
+        }
+    }
+}
+```
+To allow type parameter defaults in the same listing as const parameters we also intend to remove the ordering restriction for
+type and const parameters, allowing `struct Foo<const N: usize, T = [u32; N]> { ... }`.
+
+This feature is pretty much ready for stabilization and is currently blocked on figuring out any potential edge cases for the
+stabilization report.
+
+### `feature(generic_arg_infer)`
```

Comment:
hmm why is this feature called generic_arg_infer shouldnt it mention `const` somewhere :sweat_smile:

Reply:
because it adds a `hir::GenericArg::Infer` :laughing:

maybe `const_argument_wildcard` or something like that would be better :shrug: especially if we allow `[u8; _]` at the same time

Reply:
ah i guess it makes sense then :)

Reply:
`const_wildcard` maybe? Or `const_infer`. Ideally it would apply to all type-level constants.